Flaky test: test_regression_transport_global_queues_are_initialized_on_restart_for_services

The newly introduced test for the fix of https://github.com/raiden-network/raiden/issues/3656 seems to be flaky. Already failed once in Circle but passes locally and on the second run in Circle.

```
test_regression_transport_global_queues_are_initialized_on_restart_for_services[matrix-False-channels_per_node0-2] 

raiden_network = [<App 0dcab379>, <App 987e2cde>], number_of_nodes = 2
token_addresses = [b'\x82\x95\x03Zl\xf2\x16jW\xc9%6\x04\xec]rjH\x18\xcd']
network_wait = 10.0
user_deposit_address = b'\xc6d\x9f\x8d\x9c\xbed\x8d\x9d\xb1\xa8M\xfd\xa1\xeer\xc8J\x8e\xe3'
skip_if_not_matrix = None

    @pytest.mark.parametrize('number_of_nodes', [2])
    @pytest.mark.parametrize('channels_per_node', [CHAIN])
    def test_regression_transport_global_queues_are_initialized_on_restart_for_services(
            raiden_network,
            number_of_nodes,
            token_addresses,
            network_wait,
            user_deposit_address,
            skip_if_not_matrix,  # pylint: disable=unused-argument
    ):
        """On restarts, Raiden will restore state and publish new balance proof
        updates to the global matrix room. This test will check for regressions
        in the order of which the global queues are initialized on startup.
    
        Regression test for: https://github.com/raiden-network/raiden/issues/3656.
        """
        app0, app1 = raiden_network
    
        app0.config['services']['monitoring_enabled'] = True
    
        # Send a transfer to make sure the state has a balance proof
        # to publish to the global matrix rooms
        token_address = token_addresses[0]
        chain_state = views.state_from_app(app0)
        payment_network_id = app0.raiden.default_registry.address
        token_network_identifier = views.get_token_network_identifier_by_token_address(
            chain_state,
            payment_network_id,
            token_address,
        )
    
        amount = 10
        mediated_transfer(
            app1,
            app0,
            token_network_identifier,
            amount,
            timeout=network_wait * number_of_nodes,
        )
    
        app0.stop()
    
        transport = MatrixTransport(app0.config['transport']['matrix'])
        transport.send_async = Mock()
        transport._send_raw = Mock()
    
        old_start_transport = transport.start
    
        # Check that the queue is populated before the transport sends it and empties the queue
        def start_transport(*args, **kwargs):
            # Before restart the transport's global message queue should be initialized
            # There should be 2 messages in the global queue.
            # 1 for the PFS and the other for MS
            assert len(transport._global_send_queue) == 2
            # No other messages were sent at this point
            transport.send_async.assert_not_called()
            transport._send_raw.assert_not_called()
            old_start_transport(*args, **kwargs)
    
        transport.start = start_transport
    
        raiden_event_handler = RaidenEventHandler()
        message_handler = MessageHandler()
        app0_restart = App(
            config=app0.config,
            chain=app0.raiden.chain,
            query_start_block=0,
            default_registry=app0.raiden.default_registry,
            default_secret_registry=app0.raiden.default_secret_registry,
            default_service_registry=app0.raiden.default_service_registry,
            transport=transport,
            raiden_event_handler=raiden_event_handler,
            message_handler=message_handler,
            discovery=app0.raiden.discovery,
            user_deposit=app0.raiden.chain.user_deposit(user_deposit_address),
        )
>       app0_restart.start()

raiden/tests/integration/test_raidenservice.py:125: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
raiden/app.py:152: in start
    self.raiden.start()
raiden/raiden_service.py:539: in start
    self._start_transport(chain_state)
raiden/raiden_service.py:625: in _start_transport
    prev_auth_data=chain_state.last_transport_authdata,
raiden/tests/integration/test_raidenservice.py:104: in start_transport
    transport.send_async.assert_not_called()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_mock_self = <Mock id='140025090125608'>

    def assert_not_called(_mock_self):
        """assert that the mock was never called.
        """
        self = _mock_self
        if self.call_count != 0:
            msg = ("Expected '%s' to not have been called. Called %s times." %
                   (self._mock_name or 'mock', self.call_count))
>           raise AssertionError(msg)
E           AssertionError: Expected 'mock' to not have been called. Called 2 times.

/usr/local/lib/python3.7/unittest/mock.py:792: AssertionError
```

--- a/raiden/api/v1/encoding.py
+++ b/raiden/api/v1/encoding.py
@@ -122,7 +122,7 @@ class SecretHashField(fields.Field):
         "missing_prefix": "Not a valid hex encoded value, must be 0x prefixed.",
         "invalid_data": "Not a valid hex formated string, contains invalid characters.",
         "invalid_size": (
-            f"Not a valid secrethash, decoded value is not {SECRETHASH_LENGTH} " f"bytes long."
+            f"Not a valid secrethash, decoded value is not {SECRETHASH_LENGTH} bytes long."
         ),
     }
 

--- a/raiden/network/proxies/token_network.py
+++ b/raiden/network/proxies/token_network.py
@@ -1125,9 +1125,7 @@ class TokenNetwork:
                 raise RaidenRecoverableError(msg)
 
             if network_balance + amount_to_deposit > token_network_deposit_limit:
-                msg = (
-                    f"Deposit of {amount_to_deposit} exceeded " f"the token network deposit limit."
-                )
+                msg = f"Deposit of {amount_to_deposit} exceeded the token network deposit limit."
                 raise RaidenRecoverableError(msg)
 
             raise RaidenRecoverableError("Deposit gas estimatation failed for unknown reasons")
@@ -1914,7 +1912,7 @@ class TokenNetwork:
                     #
                     # This is a race condition that cannot be prevented,
                     # therefore it is a recoverable error.
-                    msg = "update transfer was mined after the settlement " "window."
+                    msg = "update transfer was mined after the settlement window."
                     raise RaidenRecoverableError(msg)
 
                 partner_details = self._detail_participant(

--- a/raiden/network/transport/matrix/transport.py
+++ b/raiden/network/transport/matrix/transport.py
@@ -94,7 +94,7 @@ class _RetryQueue(Runnable):
         self._notify_event = gevent.event.Event()
         self._lock = gevent.lock.Semaphore()
         super().__init__()
-        self.greenlet.name = f"RetryQueue " f"recipient:{to_checksum_address(self.receiver)}"
+        self.greenlet.name = f"RetryQueue recipient:{to_checksum_address(self.receiver)}"
 
     @property
     def log(self):

--- a/raiden/raiden_service.py
+++ b/raiden/raiden_service.py
@@ -49,14 +49,17 @@ from raiden.network.proxies.service_registry import ServiceRegistry
 from raiden.network.proxies.token_network_registry import TokenNetworkRegistry
 from raiden.network.proxies.user_deposit import UserDeposit
 from raiden.raiden_event_handler import EventHandler
-from raiden.services import update_services_from_balance_proof
+from raiden.services import (
+    update_monitoring_service_from_balance_proof,
+    update_services_from_balance_proof,
+)
 from raiden.settings import MEDIATION_FEE
 from raiden.storage import sqlite, wal
 from raiden.storage.serialization import DictSerializer, JSONSerializer
 from raiden.storage.wal import WriteAheadLog
 from raiden.tasks import AlarmTask
 from raiden.transfer import node, views
-from raiden.transfer.architecture import Event as RaidenEvent, StateChange
+from raiden.transfer.architecture import BalanceProofSignedState, Event as RaidenEvent, StateChange
 from raiden.transfer.identifiers import CanonicalIdentifier
 from raiden.transfer.mediated_transfer.events import SendLockedTransfer
 from raiden.transfer.mediated_transfer.mediation_fee import FeeScheduleState
@@ -902,7 +905,8 @@ class RaidenService(Runnable):
         assert self.wal, msg
 
         current_balance_proofs = list(
-            views.detect_balance_proof_change(
+            balance_proof
+            for balance_proof in views.detect_balance_proof_change(
                 old_state=ChainState(
                     pseudo_random_generator=chain_state.pseudo_random_generator,
                     block_number=GENESIS_BLOCK_NUMBER,
@@ -912,6 +916,8 @@ class RaidenService(Runnable):
                 ),
                 current_state=chain_state,
             )
+            # only request monitoring for own BPs
+            if isinstance(balance_proof, BalanceProofSignedState)
         )
 
         log.debug(
@@ -921,7 +927,7 @@ class RaidenService(Runnable):
         )
 
         for balance_proof in current_balance_proofs:
-            update_services_from_balance_proof(self, chain_state, balance_proof)
+            update_monitoring_service_from_balance_proof(self, chain_state, balance_proof)
 
     def _initialize_whitelists(self, chain_state: ChainState) -> None:
         """ Whitelist neighbors and mediated transfer targets on transport """

--- a/raiden/services.py
+++ b/raiden/services.py
@@ -13,7 +13,7 @@ from raiden.transfer.architecture import BalanceProofSignedState, BalanceProofUn
 from raiden.transfer.identifiers import CanonicalIdentifier
 from raiden.transfer.state import ChainState
 from raiden.utils import to_rdn
-from raiden.utils.typing import TYPE_CHECKING, Address
+from raiden.utils.typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
     from raiden.raiden_service import RaidenService
@@ -30,10 +30,7 @@ def update_services_from_balance_proof(
     send_pfs_update(raiden=raiden, canonical_identifier=balance_proof.canonical_identifier)
     if isinstance(balance_proof, BalanceProofSignedState):
         update_monitoring_service_from_balance_proof(
-            raiden=raiden,
-            chain_state=chain_state,
-            new_balance_proof=balance_proof,
-            monitoring_service_contract_address=raiden.default_msc_address,
+            raiden=raiden, chain_state=chain_state, new_balance_proof=balance_proof
         )
 
 
@@ -66,10 +63,7 @@ def send_pfs_update(
 
 
 def update_monitoring_service_from_balance_proof(
-    raiden: "RaidenService",
-    chain_state: ChainState,
-    new_balance_proof: BalanceProofSignedState,
-    monitoring_service_contract_address: Address,
+    raiden: "RaidenService", chain_state: ChainState, new_balance_proof: BalanceProofSignedState
 ) -> None:
     if raiden.config["services"]["monitoring_enabled"] is False:
         return
@@ -115,7 +109,9 @@ def update_monitoring_service_from_balance_proof(
     )
 
     monitoring_message = RequestMonitoring.from_balance_proof_signed_state(
-        new_balance_proof, MONITORING_REWARD, monitoring_service_contract_address
+        balance_proof=new_balance_proof,
+        reward_amount=MONITORING_REWARD,
+        monitoring_service_contract_address=raiden.default_msc_address,
     )
     monitoring_message.sign(raiden.signer)
     raiden.transport.send_global(constants.MONITORING_BROADCASTING_ROOM, monitoring_message)

--- a/raiden/storage/sqlite.py
+++ b/raiden/storage/sqlite.py
@@ -344,9 +344,7 @@ class SQLiteStorage:
     def write_state_snapshot(self, snapshot: str, statechange_id: StateChangeID) -> SnapshotID:
         snapshot_id = self._ulid_factory(SnapshotID).new()
 
-        query = (
-            "INSERT INTO state_snapshot (" " identifier, statechange_id, data" ") VALUES(?, ?, ?)"
-        )
+        query = "INSERT INTO state_snapshot (identifier, statechange_id, data) VALUES(?, ?, ?)"
         self.conn.execute(query, (snapshot_id, statechange_id, snapshot))
         self.maybe_commit()
 

--- a/raiden/tests/integration/api/test_restapi.py
+++ b/raiden/tests/integration/api/test_restapi.py
@@ -152,7 +152,7 @@ def test_payload_with_invalid_addresses(api_server_test_instance, rest_api_port_
     assert_response_with_error(response, HTTPStatus.BAD_REQUEST)
 
     url_without_prefix = (
-        "http://localhost:{port}/api/v1/" "channels/ea674fdde714fd979de3edf0f56aa9716b898ec8"
+        "http://localhost:{port}/api/v1/channels/ea674fdde714fd979de3edf0f56aa9716b898ec8"
     ).format(port=rest_api_port_number)
 
     request = grequests.patch(

--- a/raiden/tests/integration/network/proxies/test_secret_registry.py
+++ b/raiden/tests/integration/network/proxies/test_secret_registry.py
@@ -137,9 +137,7 @@ def test_concurrent_secret_registration(secret_registry_proxy, monkeypatch):
         def count_transactions(function_name, startgas, secrets):
             for secret in secrets:
                 count[secret] += 1
-                msg = (
-                    "All secrets must be registered, " "and they all must be registered only once"
-                )
+                msg = "All secrets must be registered, and they all must be registered only once"
                 assert count[secret] == 1, msg
 
             return transact(function_name, startgas, secrets)
@@ -178,5 +176,5 @@ def test_concurrent_secret_registration(secret_registry_proxy, monkeypatch):
 
         gevent.joinall(greenlets, raise_error=True)
 
-        msg = "All secrets must be registered, " "and they all must be registered only once"
+        msg = "All secrets must be registered, and they all must be registered only once"
         assert all(count[secret] == 1 for secret in secrets), msg

--- a/raiden/tests/integration/network/transport/test_matrix_transport.py
+++ b/raiden/tests/integration/network/transport/test_matrix_transport.py
@@ -619,10 +619,7 @@ def test_monitoring_global_messages(
     raiden_service.user_deposit.effective_balance.return_value = 100
 
     update_monitoring_service_from_balance_proof(
-        raiden=raiden_service,
-        chain_state=None,
-        new_balance_proof=balance_proof,
-        monitoring_service_contract_address=bytes([1] * 20),
+        raiden=raiden_service, chain_state=None, new_balance_proof=balance_proof
     )
     gevent.idle()
 

--- a/raiden/tests/integration/test_raidenservice.py
+++ b/raiden/tests/integration/test_raidenservice.py
@@ -10,6 +10,8 @@ from raiden.constants import (
     RoutingMode,
 )
 from raiden.message_handler import MessageHandler
+from raiden.messages.monitoring_service import RequestMonitoring
+from raiden.messages.path_finding_service import PFSCapacityUpdate, PFSFeeUpdate
 from raiden.network.transport import MatrixTransport
 from raiden.raiden_event_handler import RaidenEventHandler
 from raiden.settings import DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS
@@ -20,6 +22,7 @@ from raiden.tests.utils.network import CHAIN
 from raiden.tests.utils.transfer import transfer
 from raiden.transfer.state_change import Block
 from raiden.utils import BlockNumber
+from raiden.utils.typing import PaymentAmount, PaymentID, Type
 
 
 @pytest.mark.parametrize("number_of_nodes", [1])
@@ -59,7 +62,6 @@ def run_test_regression_filters_must_be_installed_from_confirmed_block(raiden_ne
     assert not search_for_item(app0_state_changes, Block, {"block_number": target_block_num})
 
 
-@pytest.mark.xfail(reason="flaky, see issue #3714")
 @pytest.mark.parametrize("number_of_nodes", [2])
 @pytest.mark.parametrize("channels_per_node", [CHAIN])
 @pytest.mark.parametrize(
@@ -76,8 +78,8 @@ def test_regression_transport_global_queues_are_initialized_on_restart_for_servi
     Regression test for: https://github.com/raiden-network/raiden/issues/3656.
     """
     raise_on_failure(
-        raiden_network,
-        run_test_regression_transport_global_queues_are_initialized_on_restart_for_services,
+        raiden_apps=raiden_network,
+        test_function=run_test_regression_transport_global_queues_are_initialized_on_restart_for_services,  # noqa
         raiden_network=raiden_network,
         number_of_nodes=number_of_nodes,
         token_addresses=token_addresses,
@@ -90,20 +92,20 @@ def run_test_regression_transport_global_queues_are_initialized_on_restart_for_s
     raiden_network, number_of_nodes, token_addresses, network_wait, user_deposit_address
 ):
     app0, app1 = raiden_network
-
     app0.config["services"]["monitoring_enabled"] = True
 
     # Send a transfer to make sure the state has a balance proof
     # to publish to the global matrix rooms
     token_address = token_addresses[0]
 
-    amount = 10
+    amount = PaymentAmount(10)
+    payment_id = PaymentID(23)
     transfer(
         initiator_app=app1,
         target_app=app0,
         token_address=token_address,
         amount=amount,
-        identifier=1,
+        identifier=payment_id,
         timeout=network_wait * number_of_nodes,
     )
 
@@ -118,18 +120,32 @@ def run_test_regression_transport_global_queues_are_initialized_on_restart_for_s
     # Check that the queue is populated before the transport sends it and empties the queue
     def start_transport(*args, **kwargs):
         # Before restart the transport's global message queue should be initialized
-        # There should be 2 messages in the global queue.
-        # 1 for the PFS and the other for MS
-        assert len(transport._global_send_queue) == 2
-        # No other messages were sent at this point
-        transport.send_async.assert_not_called()
-        transport._send_raw.assert_not_called()
+        # There should be 3 messages in the global queue:
+        # - A `MonitorRequest` to the MS
+        # - A `PFSCapacityUpdate`
+        # - A `PFSFeeUpdate`
+        queue_copy = transport._global_send_queue.copy()
+        queued_messages = list()
+        for _ in range(len(transport._global_send_queue)):
+            queued_messages.append(queue_copy.get())
+
+        def num_matching_queued_messages(room: str, message_type: Type) -> int:
+            return len(
+                [
+                    item
+                    for item in queued_messages
+                    if item[0] == room and type(item[1]) == message_type
+                ]
+            )
+
+        assert num_matching_queued_messages(MONITORING_BROADCASTING_ROOM, RequestMonitoring) == 1
+        assert num_matching_queued_messages(PATH_FINDING_BROADCASTING_ROOM, PFSFeeUpdate) == 1
+        assert num_matching_queued_messages(PATH_FINDING_BROADCASTING_ROOM, PFSCapacityUpdate) == 1
+
         old_start_transport(*args, **kwargs)
 
     transport.start = start_transport
 
-    raiden_event_handler = RaidenEventHandler()
-    message_handler = MessageHandler()
     app0_restart = App(
         config=app0.config,
         chain=app0.raiden.chain,
@@ -140,9 +156,9 @@ def run_test_regression_transport_global_queues_are_initialized_on_restart_for_s
         default_service_registry=app0.raiden.default_service_registry,
         default_msc_address=app0.raiden.default_msc_address,
         transport=transport,
-        raiden_event_handler=raiden_event_handler,
-        message_handler=message_handler,
-        routing_mode=RoutingMode.PRIVATE,  # only monitoring is tested here
+        raiden_event_handler=RaidenEventHandler(),
+        message_handler=MessageHandler(),
+        routing_mode=RoutingMode.PFS,  # not private mode, otherwise no PFS updates are queued
         user_deposit=app0.raiden.chain.user_deposit(user_deposit_address),
     )
     app0_restart.start()

--- a/raiden/tests/unit/test_channelstate.py
+++ b/raiden/tests/unit/test_channelstate.py
@@ -514,7 +514,7 @@ def test_channelstate_receive_lockedtransfer():
         sender=invalid_balance_proof.sender,
     )
     is_valid, _, _ = channel.handle_unlock(channel_state, invalid_unlock_state_change)
-    assert not is_valid, "Unlock message with chain_id different than the " "channel's should fail"
+    assert not is_valid, "Unlock message with chain_id different than the channel's should fail"
 
     is_valid, _, msg = channel.handle_unlock(channel_state, unlock_state_change)
     assert is_valid, msg

--- a/raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py
+++ b/raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate_regression.py
@@ -597,7 +597,5 @@ def test_regression_unavailable_nodes_must_be_properly_filtered():
     )
 
     send_transfer = search_for_item(initial_iteration.events, SendLockedTransfer, {})
-    msg = (
-        "All available routes are with unavailable nodes, therefore no send " "should be produced"
-    )
+    msg = "All available routes are with unavailable nodes, therefore no send should be produced"
     assert send_transfer is None, msg

--- a/raiden/tests/utils/mocks.py
+++ b/raiden/tests/utils/mocks.py
@@ -131,6 +131,7 @@ class MockRaidenService:
         self.default_registry = Mock()
         self.default_registry.address = factories.make_address()
         self.default_one_to_n_address = factories.make_address()
+        self.default_msc_address = factories.make_address()
 
         self.targets_to_identifiers_to_statuses = defaultdict(dict)
         self.route_to_feedback_token = {}

--- a/raiden/transfer/channel.py
+++ b/raiden/transfer/channel.py
@@ -846,7 +846,7 @@ def is_valid_unlock(
         # Unlock messages must increase the transferred_amount by lock amount,
         # otherwise the sender is trying to play the protocol and steal token.
         msg = (
-            "Invalid Unlock message. " "Balance proof's wrong locked_amount, expected: {} got: {}."
+            "Invalid Unlock message. Balance proof's wrong locked_amount, expected: {} got: {}."
         ).format(expected_locked_amount, received_balance_proof.locked_amount)
 
         result = (False, msg, None)

--- a/raiden/transfer/state.py
+++ b/raiden/transfer/state.py
@@ -513,7 +513,7 @@ class ChainState(State):
 
     def __repr__(self):
         return (
-            "ChainState(block_number={} block_hash={} networks={} " "qty_transfers={} chain_id={})"
+            "ChainState(block_number={} block_hash={} networks={} qty_transfers={} chain_id={})"
         ).format(
             self.block_number,
             to_hex(self.block_hash),

--- a/tools/check_circleci_config.py
+++ b/tools/check_circleci_config.py
@@ -43,7 +43,7 @@ def _check_workflows_align(config):
     )
     message = [
         _yellow(
-            "Mismatch in common items of workflow definitions 'raiden-default' and " "'nightly':\n"
+            "Mismatch in common items of workflow definitions 'raiden-default' and 'nightly':\n"
         )
     ]
     for line in job_diff:

--- a/tools/debugging/json-log-to-html.py
+++ b/tools/debugging/json-log-to-html.py
@@ -480,7 +480,7 @@ def render_fields(record: Record, sorted_known_fields: List[str]) -> List[str]:
             continue
         field_value = record.fields[field_name]
         colorized_value = str(colorize_value(field_value, min_luminance=0.6))
-        rendered_fields.append(f'<span class="fn">{field_name}</span>' f"=" f"{colorized_value}")
+        rendered_fields.append(f'<span class="fn">{field_name}</span> = {colorized_value}')
     return rendered_fields
 
 

